PrestaShop/ps_facetedsearch
# Ability for module to work on more pages

![qa](https://user-images.githubusercontent.com/6097524/216059434-82323276-a5bd-4163-9274-66b9c179c302.png)


| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | See below
| Type?         | new feature
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes PrestaShop/Prestashop#15137
| How to test?  | See below

# What this is about
- Customer can now filter on all pages! On manufacturer, supplier, prices drop, best sales, new products, search page.
- You now have extra options in module settings, where you assign which type of pages you want this filter template to be used on.
- Category checkboxes are now needed only if using the template on category page.
- In the filter overview, merchant can nicely see which pages use the filter template he created.
- There is a new column in the database generated filters that marks which controller it belongs to.

# How to test
- Test with classic theme.
- Absolutely no change was done on category pages. Just check that the filters show there and it shows some products. 
- Check if the new controllers show proper products, if sorting works and there are no new bugs specific to this new functionality. Check it on classic theme, there are some issues on hummingbird.
- There are some common issues in this module that we know about, when you find something, always compare to develop or ask me on Slack:
  - https://github.com/PrestaShop/PrestaShop/issues/29727 - test only with level of subcategories 1 in module settings.
  - https://github.com/PrestaShop/PrestaShop/issues/26803
  - https://github.com/PrestaShop/PrestaShop/issues/12501

# How does it look
![filter templates](https://user-images.githubusercontent.com/6097524/209024112-e86c67a0-d037-49d6-98a3-297641b421e9.jpg)
![template](https://user-images.githubusercontent.com/6097524/209024115-9f9aab9a-f35e-451f-9589-f3ca97bc4a4f.jpg)

# Changes made
### STEP 1 - adding controllers, administration, serialized filter templates and reacting properly:
- [x] Add supported controllers to module.
- [x] Add inputs and overview to admin.
- [x] Process admin data, extend serialized filter data by controller information.
- [x] Provide search interface only in case of supported controller.

### STEP 2 - Add controller column, generate filters, make proper SQL queries:
- [x] Modify "generated filters" DB structure, generate it with new data, get this data with a controller condition. Add upgrade.
- [x] Take new controllers into account when generating filters.

### STEP 3 - Use the new data when filtering:
- [x] Setup proper caching depending on the page.
- [x] Use correct filter data on the new pages. 
- [x] Add new conditions for new controllers that have joins.
- [x] Add new conditions for new controllers that don't have joins.
- [x] Check and implement sorting options on new controllers.

### STEP 4 - Checks and polishing:
- [x] Add compatibility with 1.7.6-1.7.8 and their missing queryTypes.
- [x] Verify product selection behavior on special pages.
- [x] Add tests for Search class.

### Issues on Hummingbird (already being fixed on theme side)
- Using price filter on search page breaks it, it removes `s` parameter from the URL.
- Removing filters on new products page for example removes them, but it keeps the old URL.

### Issues found on the core side
- On best sales page, the default sorting without this module is `sales, desc`, but with this module it's `name, asc`. 
- It's because in BestSalesControllerCore, there is `'product', 'name'` set on the query, which arrives to our module. 
- But `BestSalesProductSearchProvider` completely ignores it and overrides it by `'product', 'sales'`, which we don't have a clue about.
- **I fixed this in the module, so you won't notice, but should be fixed also on core side (https://github.com/PrestaShop/PrestaShop/pull/31074).**


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/PrestaShop/ps_facetedsearch/757)
